File: android/app/src/main/kotlin/games/yandex/wrap/catalog/dtos/GameDto.kt
package games.yandex.wrap.catalog.dtos

import games.yandex.wrap.catalog.models.Game
import kotlinx.serialization.Serializable

@Serializable
data class GameDto(
    val appID: Long,
    val title: String,
    val rating: Float = 0f,
    val ratingCount: Int = 0,
    val media: GameMediaDto? = null,
    val categoriesNames: List<String> = emptyList(),
    val developer: NamedDto? = null,
    val features: GameFeaturesDto? = null,
) {
    fun toDomain(): Game {
        val coverPrefix = media?.cover?.prefixUrl
        val iconPrefix = media?.icon?.prefixUrl
        return Game(
            appId = appID,
            title = title,
            rating = rating,
            ratingCount = ratingCount,
            coverUrl = coverPrefix.orEmpty(),
            iconUrl = iconPrefix ?: coverPrefix.orEmpty(),
            categories = categoriesNames,
            developer = developer?.name.orEmpty(),
            mainColor = media?.cover?.mainColor,
            iconMainColor = media?.icon?.mainColor,
            videoUrl = media?.videos?.firstOrNull()?.mp4StreamUrl,
            coverPrefixUrl = coverPrefix,
            ageRating = features?.ageRating,
        )
    }
}

File: ios/UGames/Catalog/DTOs/GameDTO.swift
import Foundation

struct GameDTO: Decodable {
    let appID: Int64
    let title: String
    let rating: Double?
    let ratingCount: Int?
    let media: GameMediaDTO?
    let categoriesNames: [String]?
    let developer: NamedDTO?
    let features: GameFeaturesDTO?

    var domain: Game {
        let coverPrefix = media?.cover?.prefixUrl
        let iconPrefix = media?.icon?.prefixUrl
        return Game(
            appId: appID,
            title: title,
            rating: rating ?? 0,
            ratingCount: ratingCount ?? 0,
            coverUrl: coverPrefix ?? "",
            iconUrl: iconPrefix ?? coverPrefix ?? "",
            categories: categoriesNames.orEmpty,
            developer: developer?.name ?? "",
            mainColor: media?.cover?.mainColor,
            iconMainColor: media?.icon?.mainColor,
            videoUrl: media?.videos.orEmpty.first?.mp4StreamUrl,
            coverPrefixUrl: coverPrefix,
            ageRating: features?.ageRating
        )
    }
}
